keps/sig-apps/0032-portable-service-definitions.md

```diff
@@ -0,0 +1,143 @@
+
+## Motivation
+
+Workload portability is commonly cited as a goal for those deploying workloads. The Kubernetes API can provide a common API for deploying workloads across varying environments enabling some level of portability. For example, a deployment can be run in a cluster running on GKE, AKS, EKS, or clusters running elsewhere.
+
+But, many applications rely on software as a service (SaaS). The reason for this is to push the operational details on to someone else who specializes in that particular service so the application developers and operators can focus on their application and business logic.
+
+Currently, service portability as declared through Kubernetes with a common API is not possible. This feature set would provide a means to solve that.
+
+### Goals
+
+* Provide a common way to request common services (e.g., MySQL)
+* Provide a common means to obtain credentials to use the service
+* Provide a common method to detect which services are available
+* Provide a system that can be implemented for the major public clouds, on-premise clusters, and local cluster (e.g., Docker for Mac)
+
+### Non-Goals
+
+* Provide an out of the box solution for every bespoke service provided by everyone
+* Replace Kubernetes Service Catalog
+
+## Proposal
+
+This is where we get down to the nitty gritty of what the proposal actually is.
+
+### User Stories
+
+#### Story 1
+
+As a user of Kubernetes, I can query the services I can declaratively reqest using a Kubernetes native API. For example, using the command `kubectl get crds` where I can see services alongside other resources that can be created.
+
+#### Story 2
+
+As a user of Kubernetes, I can declaratively request an instance of a service using a custom resource. When the service is provided the means to use that service (e.g., credentials in a secret) are provided in a common and consistent manner. The same resource and secret can be used in clusters running in different locations and the way the service is provided may be different.
+
+### Implementation Details/Notes/Constraints
+
+There are two important types of resources in this model. First, there are custom resource definitions (CRDs) for the services. Each service will have its own CRD. For example, if MySQL, PostgreSQL, and DNS are all available there will be three different CRDs that can be listed. To create an insteance of one a custom resource will be created following the API from the custom resource definition (CRD).
+
+The second type of resource is the optional secret to access the resource. If we continue the example, MySQL and PostgreSQL will have connection credentials that need to be used and injected into an appliction. In each case there will be a secret with a type specific to the service that follows a common schema. For services that do not require credential no secret is required.
```

Comment:
Can this be made more clear? I can't tell if the optional secret is an input or an output (I expected the latter?).

I also don't understand how the output secret can be both strongly typed, and represent all of the credential or other information the service needs. An example API would be very helpful.

Reply:
OK, I think I actually did understand what you were saying. I'm just not convinced it works in the abstract even though I think many specific instances will work--see the comment I just left.

Reply:
@lavalamp This won't work for everything. Especially all the bespoke custom services. But, there is a lot it will work for and it will be useful for many in those situations. We're not shooting for a perfect abstraction. We're shooting for useful where useful may be summed up as "useful = usability + utility"

Reply:
In that case perhaps a list of specific things it will work for would be helpful?

Reply:
While I won't try to come up with a definitive list, I did add some examples to the update to the KEP I'm working on.